### ISSUE TITLE
Massive memory leak in source processing application based on urql/core

### ISSUE THREAD

**Issue by rvalle** [FormidableLabs/urql#2353]
I am using latest @urlql/core , version 2.4.3 in source processing application.

The code itself consolidates millions of records from different tables, so there is big paginated query on the main table, and for each row additional queries may be triggered on other tables to gather additional information. The output is pushed into a different store.

The code is mostly asynchronous, with all queries run with .toPromise(), the queries on small encoding tables are cache-first, while the main query or queries with low probability of cache benefit run as network-only. I create 2 clients to be able to choose.

This could make a great stress test for urlq/core, I guess.

I see GBs in leaked memory as previously described in #2298 , I have tried rolling back to previous latest minors, with similar results.

I am currently profiling memory, and I find loads of System.context, Request, and Wonka.

Any idea what could be going wrong?

**Comment by rvalle** [comment]
I am ready about the cache, and looks like cache entries are invalidated only after a subsequent mutation. I am not using mutations at all.

**Comment by rvalle** [comment]
this could also be relevant, since we are running form node
```
import { fetch } from 'cross-fetch';
```
we are using this as the fetch package.

**Comment by rvalle** [comment]
@JoviDeCroock Removing the cache makes the process really slow, and then it becomes difficult for me to asses if memory allocation is comparable: less allocation but over much larger period of time. 

I removed cached queries and replaced the caching logic by LRU package, a simple in memory Last Recently Used logic.

The situation seems to have improved considerably. I will need to run the whole process (hours) to verify if caching was the only issue or not.

The allocation seems to be around the cached queries. But the objects in memory could be related also to non cached queries (Requests, Wonka)

**Comment by frederikhors** [comment]
@rvalle thank you for this work.

**Comment by JoviDeCroock** [comment]
@rvalle Hmmm, I guess the infinite retention period leveraged by the document cache could be contributing to this. An LRU mechanism could be more suited for this need as millions of records feel like a memory leak waiting to happen.

I guess this is not a front-end application either so a cache makes sense because it's mainly static data I assume? Or are you doing that on a front-end app?

**Comment by kitten** [comment]
Either way, closing this as expected behaviour. I'm assuming a change or feature could come out of this though

**Comment by rvalle** [comment]
Yes, this is not a fronted application, and my assumption that a "cache" implies "finite" memory store is probably the issue here.

**Comment by rvalle** [comment]
@kitten @JoviDeCroock FYI cache is not the issue.

- I disabled every processing and started focusing on the main query loop, the cursor over the main table, and seems to leak.
- I disabled the cursor and just asked for the first page, as promise, and still seems to leak.

I am starting to think that just running a query leaks memory.  

I will try to generate a minimal concept. Simple query in an infinite loop.
